File: backend/src/api/routers/events.py
from typing import Annotated, Literal, cast
from uuid import UUID

from fastapi import APIRouter, Query
from pydantic import AwareDatetime

from api.dependencies import Authenticated, Events
from api.responses import READ_ERROR_CODES, WRITE_ERROR_CODES, error_responses
from events.schemas import (
    EventInput,
    EventMemberInput,
    EventMergeInput,
    EventPage,
    EventRevisionView,
    EventSplitInput,
    EventTrendView,
    EventView,
    TrendBucketHours,
)
from notifications.schemas import (
    TrendAlertEvaluationView,
    TrendAlertRuleInput,
    TrendAlertRuleUpdate,
    TrendAlertRuleView,
)

router = APIRouter(prefix="/events", tags=["events"])


@router.get(
    "",
    response_model=EventPage,
    responses=error_responses(*READ_ERROR_CODES),
    operation_id="listEvents",
)
def events(
    service: Events,
    owner: Authenticated,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    cursor: UUID | None = None,
) -> EventPage:
    return service.events(limit, cursor)


@router.post(
    "",
    response_model=EventView,
    status_code=201,
    responses=error_responses(*WRITE_ERROR_CODES),
    operation_id="createEvent",
)
def create_event(data: EventInput, service: Events, owner: Authenticated) -> EventView:
    return service.create(data)


@router.get(
    "/{identity}",
    response_model=EventView,
    responses=error_responses(*READ_ERROR_CODES, 404),
    operation_id="getEvent",
)
def get_event(identity: UUID, service: Events, owner: Authenticated) -> EventView:
    return service.event(identity)


@router.post(
    "/{identity}/members",
    response_model=EventView,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 409),
    operation_id="addEventMember",
)
def add_event_member(
    identity: UUID, data: EventMemberInput, service: Events, owner: Authenticated
) -> EventView:
    return service.add_member(identity, data)


@router.delete(
    "/{identity}/members/{content_id}",
    response_model=EventView,
    responses=error_responses(*WRITE_ERROR_CODES, 404),
    operation_id="removeEventMember",
)
def remove_event_member(
    identity: UUID, content_id: UUID, service: Events, owner: Authenticated
) -> EventView:
    return service.remove_member(identity, content_id)


@router.get(
    "/{identity}/revisions",
    response_model=list[EventRevisionView],
    responses=error_responses(*READ_ERROR_CODES, 404),
    operation_id="listEventRevisions",
)
def event_revisions(
    identity: UUID, service: Events, owner: Authenticated
) -> list[EventRevisionView]:
    return service.revisions(identity)


@router.get(
    "/{identity}/trends",
    response_model=EventTrendView,
    responses=error_responses(*READ_ERROR_CODES, 404, 422),
    operation_id="getEventTrends",
)
def event_trends(
    identity: UUID,
    service: Events,
    owner: Authenticated,
    since: Annotated[AwareDatetime, Query()],
    until: Annotated[AwareDatetime, Query()],
    bucket_hours: TrendBucketHours = TrendBucketHours.daily,
) -> EventTrendView:
    return service.trends(identity, since, until, cast(Literal[1, 6, 24], int(bucket_hours)))


@router.get(
    "/{identity}/trend-alert-rules",
    response_model=list[TrendAlertRuleView],
    responses=error_responses(*READ_ERROR_CODES, 404),
    operation_id="listEventTrendAlertRules",
)
def event_trend_alert_rules(
    identity: UUID, service: Events, owner: Authenticated
) -> list[TrendAlertRuleView]:
    return service.trend_alert_rules(identity)


@router.post(
    "/{identity}/trend-alert-rules",
    response_model=TrendAlertRuleView,
    status_code=201,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 409),
    operation_id="createEventTrendAlertRule",
)
def create_event_trend_alert_rule(
    identity: UUID,
    data: TrendAlertRuleInput,
    service: Events,
    owner: Authenticated,
) -> TrendAlertRuleView:
    return service.create_trend_alert_rule(identity, data)


@router.patch(
    "/{identity}/trend-alert-rules/{rule_id}",
    response_model=TrendAlertRuleView,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 409),
    operation_id="updateEventTrendAlertRule",
)
def update_event_trend_alert_rule(
    identity: UUID,
    rule_id: UUID,
    data: TrendAlertRuleUpdate,
    service: Events,
    owner: Authenticated,
) -> TrendAlertRuleView:
    return service.update_trend_alert_rule(identity, rule_id, data)


@router.post(
    "/{identity}/trend-alerts/evaluate",
    response_model=TrendAlertEvaluationView,
    responses=error_responses(*WRITE_ERROR_CODES, 404),
    operation_id="evaluateEventTrendAlerts",
)
def evaluate_event_trend_alerts(
    identity: UUID, service: Events, owner: Authenticated
) -> TrendAlertEvaluationView:
    return service.evaluate_trend_alerts(identity)


@router.post(
    "/{identity}/merge",
    response_model=EventView,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 409),
    operation_id="mergeEvent",
)
def merge_event(
    identity: UUID, data: EventMergeInput, service: Events, owner: Authenticated
) -> EventView:
    return service.merge(identity, data)


@router.post(
    "/{identity}/split",
    response_model=EventView,
    status_code=201,
    responses=error_responses(*WRITE_ERROR_CODES, 404, 409),
    operation_id="splitEvent",
)
def split_event(
    identity: UUID, data: EventSplitInput, service: Events, owner: Authenticated
) -> EventView:
    return service.split(identity, data)

File: frontend/src/api/events.ts
// @ts-ignore
/* eslint-disable */
import { request, type RequestOptions } from "../request";

/** Events GET /api/v1/events */
export async function listEvents(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listEventsParams,
  options?: RequestOptions
) {
  return request<API.EventPage>("/api/v1/events", {
    method: "GET",
    params: {
      // limit has a default value: 20
      limit: "20",
      ...params,
    },
    ...(options || {}),
  });
}

/** Create Event POST /api/v1/events */
export async function createEvent(
  body: API.EventInput,
  options?: RequestOptions
) {
  return request<API.EventView>("/api/v1/events", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** Get Event GET /api/v1/events/${param0} */
export async function getEvent(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getEventParams,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.EventView>(`/api/v1/events/${param0}`, {
    method: "GET",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** Add Event Member POST /api/v1/events/${param0}/members */
export async function addEventMember(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.addEventMemberParams,
  body: API.EventMemberInput,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.EventView>(`/api/v1/events/${param0}/members`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    params: { ...queryParams },
    data: body,
    ...(options || {}),
  });
}

/** Remove Event Member DELETE /api/v1/events/${param0}/members/${param1} */
export async function removeEventMember(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.removeEventMemberParams,
  options?: RequestOptions
) {
  const { identity: param0, content_id: param1, ...queryParams } = params;
  return request<API.EventView>(`/api/v1/events/${param0}/members/${param1}`, {
    method: "DELETE",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** Merge Event POST /api/v1/events/${param0}/merge */
export async function mergeEvent(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.mergeEventParams,
  body: API.EventMergeInput,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.EventView>(`/api/v1/events/${param0}/merge`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    params: { ...queryParams },
    data: body,
    ...(options || {}),
  });
}

/** Event Revisions GET /api/v1/events/${param0}/revisions */
export async function listEventRevisions(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listEventRevisionsParams,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.EventRevisionView[]>(
    `/api/v1/events/${param0}/revisions`,
    {
      method: "GET",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** Split Event POST /api/v1/events/${param0}/split */
export async function splitEvent(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.splitEventParams,
  body: API.EventSplitInput,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.EventView>(`/api/v1/events/${param0}/split`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    params: { ...queryParams },
    data: body,
    ...(options || {}),
  });
}

/** Event Trends GET /api/v1/events/${param0}/trends */
export async function getEventTrends(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getEventTrendsParams,
  options?: RequestOptions
) {
  const { identity: param0, ...queryParams } = params;
  return request<API.EventTrendView>(`/api/v1/events/${param0}/trends`, {
    method: "GET",
    params: {
      // bucket_hours has a default value: 24
      bucket_hours: "24",
      ...queryParams,
    },
    ...(options || {}),
  });
}
